batch/batch/front_end/front_end.py

```diff
@@ -1414,22 +1417,45 @@ async def delete(tx):
 ''',
             (billing_project, user))
     await delete()  # pylint: disable=no-value-for-parameter
-    set_message(session, f'Removed user {user} from billing project {billing_project}.', 'info')
-    return web.HTTPFound(deploy_config.external_url('batch', '/billing_projects'))
 
 
-@routes.post('/billing_projects/{billing_project}/users/add')
-@prom_async_time(REQUEST_TIME_POST_BILLING_PROJECT_ADD_USER_UI)
+@routes.post('/billing_projects/{billing_project}/users/{user}/remove')
+@prom_async_time(REQUEST_TIME_POST_BILLING_PROJECT_REMOVE_USER_UI)
 @check_csrf_token
 @web_authenticated_developers_only(redirect=False)
-async def post_billing_projects_add_user(request, userdata):  # pylint: disable=unused-argument
+async def post_billing_projects_remove_user(request, userdata):  # pylint: disable=unused-argument
     db = request.app['db']
-    post = await request.post()
-    user = post['user']
     billing_project = request.match_info['billing_project']
+    user = request.match_info['user']
 
     session = await aiohttp_session.get_session(request)
+    try:
+        await _remove_user_from_billing_project(db, billing_project, user)
+        set_message(session, f'Removed user {user} from billing project {billing_project}.', 'info')
+    except KeyError as e:
+        set_message(session, f'No such billing project {str(e)}.', 'error')
+    except ValueError as e:
+        set_message(session, f'User {str(e)} is not member of billing project {billing_project}.', 'info')
```

Comment:
Using standard Python exceptions opens us up to the risk that a bug will be misinterpreted as a normal error instead of an ERROR log.

The control flow of that method is a bit hard to fit into the model you want. You'll either need to create a couple classes to represent your return/exceptional values, provide functions to execute in the different cases, or just break that method into smaller pieces that can be called independently (in this case you push the transaction into each of these route handlers).

Reply:
Which method? I think this looks fine if I throw batch specific errors and keep the rest the same, right? It actually probably makes sense to define an API wrapper and a UI wrapper that does the right thing for each of these functions, since the errors across these route handlers overlap a fair amount. so something like 
```
handle_batch_exceptions_ui(_remove_user_from_billing_project, ...)
handle_batch_exceptions_api(_remove_user_from_billing_project, ...)
```
to define a "standard" set of responses for various exceptions thrown so we don't have to worry about unifying response codes/messages across disparate locations.